odoo/models.py

```diff
@@ -6287,15 +6277,49 @@ def modified(self, fnames, create=False, before=False):
         #  - mark H to recompute on inverse(X, records),
         #  - mark I to recompute on inverse(W, inverse(X, records)),
         #  - mark J to recompute on inverse(Y, records).
-        if len(fnames) == 1:
-            tree = self.pool.field_triggers.get(self._fields[next(iter(fnames))])
-        else:
-            # merge dependency trees to evaluate all triggers at once
-            tree = {}
-            for fname in fnames:
-                node = self.pool.field_triggers.get(self._fields[fname])
-                if node:
-                    trigger_tree_merge(tree, node)
+        fields = [self._fields[fname] for fname in fnames]
+        field_triggers = self.pool.field_triggers
+        trees = [field_triggers[field] for field in fields if field in field_triggers]
+
+        if not trees:
+            return
+
+        cache = self.env.cache
+
+        def merge_trees(trees: list) -> dict:
```

Comment:
I suggest to move that function out of the method.
All it needs is an extra parameter that allows to filter out some fields.
```py
def merge_trees(trees: list, select=bool) -> dict:
    """ Merge the given trigger trees into a final tree.

    :param select: the function is applied on every field in the tree, and the
        field is discarded if the function returns ``False``
    """
```
The function will then be called as
```py
tree = merge_trees(trees, select=lambda field: (field.compute and field.store or cache.contains_field(field)))
```

Reply:
Thx, It's actually much nicer!